File: android/src/main/kotlin/com/michaellazebny/jyphoon/jc/platformViews/CallView.kt
package com.michaellazebny.jyphoon.jc.platformViews

import android.content.Context
import android.view.View
import android.view.ViewGroup
import com.juphoon.cloud.JCMediaDevice
import com.michaellazebny.jyphoon.jc.SelfViewApi
import com.michaellazebny.jyphoon.jc.api.ViewCanvasApi
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class CallView(messenger: BinaryMessenger, canvasApi: ViewCanvasApi) : PlatformView, SelfViewApi {
    private var view: View?

    init {
        when (canvasApi) {
            is ViewCanvasApi.SelfViewCanvasApi -> {
                SelfViewApi.setUp(messenger, this)
            }
            is ViewCanvasApi.CompanionViewCanvasApi -> {
                SelfViewApi.setUp(messenger, this)
            }
        }
        view = canvasApi.startVideo(JCMediaDevice.RENDER_FULL_SCREEN)
    }

    override fun getView() = view

    override fun dispose() {
        // remove a view of its parent
        val parent = view?.parent as ViewGroup?
        parent?.removeView(view)
    }

    override fun setSelfFrame(width: Double, height: Double) {
        view?.layoutParams?.width = ViewHelper.convertDpToPixel(width).toInt()
        view?.layoutParams?.height = ViewHelper.convertDpToPixel(height).toInt()
    }
}

class CallViewFactory(private val messenger: BinaryMessenger, private val canvasApi: ViewCanvasApi) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        return CallView(messenger, canvasApi)
    }
}

File: ios/Classes/platformViews/CallView.swift
import Flutter
import UIKit
import JCSDKOC

class CallViewFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger
    private var api: ViewCanvasApi
    
    public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
          return FlutterStandardMessageCodec.sharedInstance()
    }

    init(messenger: FlutterBinaryMessenger, api: ViewCanvasApi) {
        self.messenger = messenger
        self.api = api
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return CallView(binaryMessenger: messenger, api: api)
    }
}

class CallView: NSObject, FlutterPlatformView, CompanionViewApi, SelfViewApi {
    func setSelfFrame(width: Double, height: Double) {
        _canvas?.videoView?.frame = CGRect(x: 0, y: 0, width: width, height: height)
    }
    
    func setCompanionFrame(width: Double, height: Double) {
        _canvas?.videoView?.frame = CGRect(x: 0, y: 0, width: width, height: height)
    }
    
    private var _view: UIView
    private var _canvas: JCMediaDeviceVideoCanvas?

    init(
        binaryMessenger messenger: FlutterBinaryMessenger?,
        api: ViewCanvasApi
    ) {
        _view = UIView()
        super.init()

        if (api is CompanionViewCanvasApi) {
            CompanionViewApiSetup.setUp(binaryMessenger: messenger!, api: self)
        } else if (api is SelfViewCanvasApi) {
            SelfViewApiSetup.setUp(binaryMessenger: messenger!, api: self)
        }
        
        _canvas = api.startVideo(mode: .fullScreen)
        _view.addSubview(_canvas!.videoView)
    }

    func view() -> UIView {
        return _view
    }
}
